hyperbolic-rs/preprocess.py

```diff
@@ -104,63 +124,120 @@ def create_splits(samples, do_random=False, seed=42):
             continue
-        src_dist = item_item_distances_dict[src_iid]
-        for dst_iid, distance in src_dist.items():
-            if src_iid != dst_iid and dst_iid in iid2id:
-                dst_index = iid2id[dst_iid]
-                distance_matrix[src_index, dst_index] = max(distance, min_distance)
-    if do_sparse:
-        return sparse.csr_matrix(distance_matrix)
-    return distance_matrix
+        src_id = iid2id[src_iid]
+        sorted_dists = sorted(dists, key=lambda t: t[1])
+        added = 0
+        for dst_iid, cos_dist in sorted_dists:
+            if dst_iid not in iid2id or cos_dist > 0.3:
+                continue
+            dst_id = iid2id[dst_iid]
+            if cos_dist <= 0.1:
+                triplets.add((src_id, Relations.SEM_HIGH_SIM.value, dst_id))
+            elif 0.2 >= cos_dist > 0.1:
+                triplets.add((src_id, Relations.SEM_MEDIUM_SIM.value, dst_id))
+            else:   # 0.3 >= cos_dist > 0.2
+                triplets.add((src_id, Relations.SEM_LOW_SIM.value, dst_id))
+            added += 1
+            if added >= top_k:
+                break
+    return list(triplets)
+
+
+def add_to_train_split(data, triplets):
+    """
+    Adds the given list of triplets to the training data
+    :param data: splits with train data
+    :param triplets: list of triplets
+    """
+    train = data["train"]
+    triplets = np.array(triplets).astype('int64')
+    data["train"] = np.concatenate((train, triplets), axis=0)
+
+
+def get_co_triplets(item_metas, get_aspect_func, iid2id, relation_id):
+    """
+    Creates triplets based on co_buy or co_view relations taken from metadata
+    :param item_metas: list of amazon metadata objects
+    :param get_aspect_func: a function that extract either co_buy or co_view data
+    :param iid2id: dict of item ids
+    :param relation_id: relation index
+    :return: list of triplets
+    """
+    triplets = set()
+    for item in item_metas:
+        head_id = iid2id[item.id]
+        for co_rel_id in get_aspect_func(item):
+            if co_rel_id in iid2id:
+                tail_id = iid2id[co_rel_id]
+                triplets.add((head_id, relation_id, tail_id))
+    return list(triplets)
+
+
+def get_category_triplets(item_metas, cat2id, iid2id, relation_id):
+    """
+    Builds triplets based on categorical labels.
+    For each item: (item, has_category, category)
+
+    :param item_metas: list of amazon metadata objects
+    :param cat2id: dict of category ids
+    :param iid2id: dict of item ids
+    :param relation_id: relation index for has_category relation
+    :return: list of triplets
+    """
+    triplets = set()
+    for it_meta in item_metas:
+        for cat in it_meta.categories:
+            item_id = iid2id[it_meta.id]
+            cat_id = cat2id[cat]
+            triplets.add((item_id, relation_id, cat_id))
+    return list(triplets)
+
+
+def get_brand_triplets(item_metas, brand2id, iid2id, relation_id):
```

Comment:
Just a cosmetic suggestion: If you feel generous, you could move a few of those into its own file / module.
There are currently  a lot of unrelated things going on in here.

Reply:
I moved the logic for the extra amazon relations into a separate file.
Thanks!